networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/peers/PeerChainValidator.java

```diff
@@ -0,0 +1,197 @@
+import tech.pegasys.artemis.datastructures.networking.libp2p.rpc.GoodbyeMessage;
+import tech.pegasys.artemis.datastructures.state.Checkpoint;
+import tech.pegasys.artemis.networking.eth2.peers.Eth2Peer.StatusData;
+import tech.pegasys.artemis.storage.ChainStorageClient;
+import tech.pegasys.artemis.storage.HistoricalChainData;
+import tech.pegasys.artemis.util.SSZTypes.Bytes4;
+
+public class PeerChainValidator {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final ChainStorageClient storageClient;
+  private final HistoricalChainData historicalChainData;
+  private final Eth2Peer peer;
+  private final AtomicBoolean hasRun = new AtomicBoolean(false);
+
+  private PeerChainValidator(
+      final ChainStorageClient storageClient,
+      final HistoricalChainData historicalChainData,
+      final Eth2Peer peer) {
+    this.storageClient = storageClient;
+    this.historicalChainData = historicalChainData;
+    this.peer = peer;
+  }
+
+  public static PeerChainValidator create(
+      final ChainStorageClient storageClient,
+      final HistoricalChainData historicalChainData,
+      final Eth2Peer peer) {
+    return new PeerChainValidator(storageClient, historicalChainData, peer);
+  }
+
+  public void run() {
+    if (hasRun.compareAndSet(false, true)) {
+      executeCheck();
+    }
+  }
+
+  private void executeCheck() {
+    LOG.trace("Validate chain of peer: {}", peer);
+    checkRemoteChain()
+        .whenComplete(
+            (isValid, error) -> {
+              if (error != null) {
+                LOG.debug("Unable to validate peer's chain, disconnecting: " + peer, error);
+                peer.sendGoodbye(GoodbyeMessage.REASON_UNABLE_TO_VERIFY_NETWORK);
+                return;
+              } else if (!isValid) {
+                // We are not on the same chain
+                LOG.trace("Disconnecting peer on different chain: {}", peer);
+                peer.sendGoodbye(GoodbyeMessage.REASON_IRRELEVANT_NETWORK);
+                return;
+              }
+              LOG.trace("Validated peer's chain: {}", peer);
+              peer.markChainValidated();
+            });
+  }
+
+  private CompletableFuture<Boolean> checkRemoteChain() {
+    checkState(peer.hasStatus(), "Peer must have an associated status.");
+    final StatusData peerStatus = peer.getStatus();
+
+    final CompletableFuture<Boolean> isChainValid = new CompletableFuture<>();
+
+    // Check fork compatibility
+    Bytes4 expectedFork = storageClient.getForkAtSlot(peerStatus.getHeadSlot());
+    if (!Objects.equals(expectedFork, peerStatus.getHeadForkVersion())) {
+      LOG.trace(
+          "Peer's fork ({}) differs from our fork ({}): {}",
+          peerStatus.getHeadForkVersion(),
+          expectedFork,
+          peer);
+      isChainValid.complete(false);
+      return isChainValid;
+    }
+
+    // If we haven't reached genesis, accept our peer at this point
+    if (storageClient.isPreGenesis()) {
```

Comment:
Created a follow-up ticket to handle this case more reliably: https://pegasys1.atlassian.net/browse/BC-180